utils/frame/sub-du/src/main.rs

```diff
@@ -0,0 +1,271 @@
+			Self::Map(bytes, count) => {
+				write!(f, "Map({} bytes, {} keys)", bytes.separated_string(), count)
+			},
+		}
+	}
+}
+
+impl Default for StorageItem {
+	fn default() -> Self {
+		Self::Value(0)
+	}
+}
+
+#[derive(Debug, Clone, Default)]
+struct Storage {
+	pub name: String,
+	pub size: usize,
+	pub item: StorageItem,
+}
+
+impl std::fmt::Display for Storage {
+	fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+		let item_style = Style::new().italic();
+		write!(f, "{} => {}", item_style.paint(self.name.clone()), self.item)
+	}
+}
+
+impl Storage {
+	fn new(name: String, item: StorageItem) -> Self {
+		let size = match item {
+			StorageItem::Value(s) => s,
+			StorageItem::Map(s, _) => s,
+		};
+		Self { name, item, size }
+	}
+}
+
+#[derive(Debug, StructOpt)]
+#[structopt(
+	name = "sub-du",
+	about = "a du-like tool that prints the map of storage usage of a substrate chain"
+)]
+struct Opt {
+	/// The block number at which the scrap should happen. Use only the hex value, no need for a
+	/// `0x` prefix.
+	#[structopt(long)]
+	at: Option<Hash>,
+
+	/// The node to connect to.
+	#[structopt(long, default_value = "wss://rpc.polkadot.io:443")]
+	uri: String,
+
+	/// If true, intermediate values will be printed.
+	#[structopt(long, short)]
+	progress: bool,
+
+	/// Weather to scrape all pairs or just the size of them.
+	///
+	/// If enabled, the command might take longer but then the number of keys in each map is also
+	/// scraped.
+	///
+	/// # Warning
+	///
+	/// This uses an unsafe RPC call and can only be used if the target node allows it.
+	#[structopt(long, short)]
+	scrape_pairs: bool,
+}
+
+/// create key prefix for a module as vec bytes. Basically twox128 hash of the given values.
+pub fn pallet_prefix_raw(module: &[u8], storage: &[u8]) -> Vec<u8> {
+	let module_key = twox_128(module);
+	let storage_key = twox_128(storage);
+	let mut final_key = Vec::with_capacity(module_key.len() + storage_key.len());
+	final_key.extend_from_slice(&module_key);
+	final_key.extend_from_slice(&storage_key);
+	final_key
+}
+
+type Block = RawBlock<ExtrinsicWrapper<Hash>>;
+
+#[tokio::main]
+async fn main() -> () {
+	tracing_subscriber::fmt::try_init().unwrap();
+	let opt = Opt::from_args();
+
+	let mut cfg = OnlineConfig::default();
+	cfg.transport = Transport::Uri(opt.uri);
+	cfg.transport.map_uri().await.unwrap();
+
+	// connect to a node.
+	let ext = remote_externalities::Builder::<Block>::new().mode(Mode::Online(cfg));
+	let mut modules: Vec<Pallet> = vec![];
+
+	// potentially replace head with the given hash
+	let head = ext.rpc_get_head().await.unwrap();
+	let at = opt.at.or(Some(head));
+
+	let rpc_client = ext.as_online().rpc_client();
+
+	let runtime = rpc_client.runtime_version(at).await.unwrap();
+
+	println!("Scraping at block {:?} of {}({})", at, runtime.spec_name, runtime.spec_version,);
+
+	let raw_metadata = rpc_client.metadata(at).await.unwrap();
```

Comment:
where is this `fn metadata` coming from? 🤔

Reply:
it's from the `substrate_rpc_client::StateApi` i.e, https://github.com/paritytech/substrate/blob/master/client/rpc-api/src/state/mod.rs#L79

Reply:
I see, then using the RPC stuff directly is almost an identical code, you just use this client: https://github.com/paritytech/substrate/blob/4b4a1e91ef8048636e704cd60d0dcf89572dd88b/utils/frame/rpc/client/src/lib.rs#L59

Reply:
You can also consider parallelizing this like my #12537

Reply:
> I see, then using the RPC stuff directly is almost an identical code, you just use this client:
> 
> https://github.com/paritytech/substrate/blob/4b4a1e91ef8048636e704cd60d0dcf89572dd88b/utils/frame/rpc/client/src/lib.rs#L59

Makes sense, indeed remote_externalities use seems a bit hacky, especially if it does not yield all that many benefits.